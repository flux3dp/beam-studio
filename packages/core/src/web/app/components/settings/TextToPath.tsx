/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { useMemo } from 'react';

import onOffOptionFactory from 'app/components/settings/onOffOptionFactory';
import SelectControl from 'app/components/settings/SelectControl';
import useI18n from 'helpers/useI18n';

interface Props {
  getBeamboxPreferenceEditingValue: (key: string) => any;
  updateBeamboxPreferenceChange: (item_key: string, newVal: any) => void;
}

function TextToPath({
  getBeamboxPreferenceEditingValue,
  updateBeamboxPreferenceChange,
}: Props): JSX.Element {
  const lang = useI18n();

  const isFontSubstitutionOn = getBeamboxPreferenceEditingValue('font-substitute') !== false;
  const fontSubstituteOptions = onOffOptionFactory(isFontSubstitutionOn, { lang });
  const defaultFontConvert = getBeamboxPreferenceEditingValue('font-convert') || '2.0';
  const defaultLaserModuleOptions = useMemo(
    () => [
      { value: '1.0', label: '1.0', selected: defaultFontConvert === '1.0' },
      { value: '2.0', label: '2.0', selected: defaultFontConvert === '2.0' },
    ],
    [defaultFontConvert]
  );
  return (
    <>
      <div className="subtitle">{lang.settings.groups.text_to_path}</div>
      <SelectControl
        label={lang.settings.font_substitute}
        url={lang.settings.help_center_urls.font_substitute}
        id="font-substitue"
        options={fontSubstituteOptions}
        onChange={(e) => updateBeamboxPreferenceChange('font-substitute', e.target.value)}
      />
      <SelectControl
        label={lang.settings.font_convert}
        url={lang.settings.help_center_urls.font_convert}
        id="font-convert"
        options={defaultLaserModuleOptions}
        onChange={(e) => updateBeamboxPreferenceChange('font-convert', e.target.value)}
      />
    </>
  );
}

export default TextToPath;
