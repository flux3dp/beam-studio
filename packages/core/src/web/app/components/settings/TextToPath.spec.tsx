import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import { create } from 'zustand';

const mockGetPreference = jest.fn();
const mockSetPreference = jest.fn();

const useSettingStore = create(() => ({
  getPreference: mockGetPreference,
  setPreference: mockSetPreference,
}));

jest.mock('@core/app/pages/Settings/useSettingStore', () => ({
  useSettingStore,
}));

jest.mock('./components/SettingSelect');

import TextToPath from './TextToPath';

test('should render correctly', () => {
  mockGetPreference.mockImplementation((key: string) => (key === 'font-substitute' ? true : '2.0'));

  const { container } = render(
    <TextToPath
      options={
        [
          { label: 'On', value: true },
          { label: 'Off', value: false },
        ] as any
      }
    />,
  );

  expect(container).toMatchSnapshot();

  const controls = container.querySelectorAll('.select-control');

  fireEvent.change(controls[0], { target: { value: false } });
  expect(mockSetPreference).toHaveBeenCalledTimes(1);
  expect(mockSetPreference).toHaveBeenNthCalledWith(1, 'font-substitute', false);

  fireEvent.change(controls[1], { target: { value: '2.0' } });
  expect(mockSetPreference).toHaveBeenCalledTimes(2);
  expect(mockSetPreference).toHaveBeenNthCalledWith(2, 'font-convert', '2.0');
});
