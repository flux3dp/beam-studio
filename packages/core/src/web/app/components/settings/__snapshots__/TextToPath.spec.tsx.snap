// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div
    class="subtitle"
  >
    Text
  </div>
  <div>
    mock-select-control id:
    font-substitue
    label:
    Substitute Unsupported Characters
    url:
    https://support.flux3dp.com/hc/en-us/articles/360004496575
    options:
    [{"value":"TRUE","label":"On","selected":true},{"value":"FALSE","label":"Off","selected":false}]
    <input
      class="select-control"
    />
  </div>
  <div>
    mock-select-control id:
    font-convert
    label:
    Convert Text to Path
    url:
    https://support.flux3dp.com/hc/Convert-Text-to-Path
    options:
    [{"value":"1.0","label":"1.0","selected":true},{"value":"2.0","label":"2.0","selected":false}]
    <input
      class="select-control"
    />
  </div>
</div>
`;
